File: src/com/epri/dss/general/impl/DSSObjectImpl.java
/**
 * Copyright (C) 2009 Electric Power Research Institute, Inc.
 * Copyright (C) 2011 Richard Lincoln
 *
 * All rights reserved.
 */

package com.epri.dss.general.impl;

import java.io.File;
import java.io.FileOutputStream;
import java.io.PrintStream;

import com.epri.dss.common.DSSClass;
import com.epri.dss.common.impl.DSSClassImpl;
import com.epri.dss.general.DSSObject;

public class DSSObjectImpl extends NamedObjectImpl implements DSSObject {

	protected int PropSeqCount;

	protected String[] PropertyValue;

	protected int[] PrpSequence;

	/* PD, PC, Monitor, CondCode, etc. */
	protected int DSSObjType;

	protected DSSClass ParentClass;

	/* Index into the class collection list */
	protected int ClassIndex;

	protected boolean HasBeenSaved;

	/* General purpose Flag for each object  don't assume inited */
	protected boolean Flag;


	public DSSObjectImpl(DSSClass ParClass) {
		super(ParClass.getName());

		DSSObjType = 0;
		PropSeqCount = 0;
		ParentClass = ParClass;
		PropertyValue = new String[ParentClass.getNumProperties()];

		// init'd to zero when allocated
		PrpSequence = new int[ParentClass.getNumProperties()];

		HasBeenSaved = false;
	}

	protected void finalize() throws Throwable {
		for (int i = 0; i < ParentClass.getNumProperties(); i++) {
			PropertyValue[i] = "";
		}
		PropertyValue = new String[0];
		PrpSequence = new int[0];

		super.finalize();
	}

	public int getDSSObjType() {
		return DSSObjType;
	}

	public void setDSSObjType(int dSSObjType) {
		DSSObjType = dSSObjType;
	}

	public DSSClass getParentClass() {
		return ParentClass;
	}

	public void setParentClass(DSSClass parentClass) {
		ParentClass = parentClass;
	}

	public int getClassIndex() {
		return ClassIndex;
	}

	public void setClassIndex(int classIndex) {
		ClassIndex = classIndex;
	}

	public boolean isHasBeenSaved() {
		return HasBeenSaved;
	}

	public void setHasBeenSaved(boolean hasBeenSaved) {
		HasBeenSaved = hasBeenSaved;
	}

	public boolean isFlag() {
		return Flag;
	}

	public void setFlag(boolean flag) {
		Flag = flag;
	}

	/* Use DSSClass.PropertyIndex to get index by name */
	public String getPropertyValue(int Index) {
		// Default Behavior for all DSS Objects
		return this.PropertyValue[Index];
	}

	public void setPropertyValue(int Index, String Value) {
		PropertyValue[Index] = Value;

		// Keep track of the order in which this property was
		// accessed for Save Command
		PropSeqCount += 1;
		PrpSequence[Index] = PropSeqCount;
	}

	public String getName() {
		return getLocalName();
	}

	public void setName(String Value) {
		// If renamed, then let someone know so hash list can be updated
		if (getLocalName().length() > 0)
			ParentClass.setElementNamesOutOfSynch(true);

		setLocalName(Value);
	}

	/*
	 * Find next larger property sequence number
	 * return 0 if none found
	 */
	protected int GetNextPropertySet(int idx) {
		int Smallest = 9999999; // some big number
		int Result = 0;

		if (idx > 0)
			idx = PrpSequence[idx];

		for (int i = 0; i < ParentClass.getNumProperties(); i++) {
			if (PrpSequence[i] > idx) {
				if (PrpSequence[i] < Smallest) {
					Smallest = PrpSequence[i];
					   Result = i;
				}
			}
		}

		return Result;
	}

	/* Allow Calls to edit from object itself */
	public int edit() {
		ParentClass.setActive(ClassIndex);
		return 0;//super.edit();
	}

	public void initPropertyValues(int ArrayOffset) {
		this.PropertyValue[ArrayOffset] = "";

		// Clear propertySequence Array after initialization
		clearPropSeqArray();
	}

	public void dumpProperties(PrintStream F, boolean Complete) {
		F.println();
		F.println("New " + getDSSClassName() + '.' + getName());
	}

	public void saveWrite(PrintStream F) {
		/* Write only properties that were explicitly set in the
		final order they were actually set */
		int iProp = GetNextPropertySet(0); // Works on ActiveDSSObject
		while (iProp > 0) {
			DSSClass pc = ParentClass;
//			F.print(' ' + pc.getPropertyName(pc.getRevPropertyIdxMap()[iProp]));
//			F.print('=' + Utilities.CheckForBlanks(PropertyValue[iProp]));
			iProp = GetNextPropertySet(iProp);
		}
	}

	public void clearPropSeqArray() {
		PropSeqCount = 0;
		for (int i = 0; i < ParentClass.getNumProperties(); i++) {
			PrpSequence[i] = 0;
		}
	}

}

File: pydss/general/DSSObjectImpl.py
from dss.common.impl.Utilities import (Utilities,)
from dss.general.impl.NamedObjectImpl import (NamedObjectImpl,)
from dss.general.DSSObject import (DSSObject,)
# from java.io.PrintWriter import (PrintWriter,)


class DSSObjectImpl(NamedObjectImpl, DSSObject):
    propSeqCount = None
    propertyValue = None
    prpSequence = None
    objType = None
    parentClass = None
    classIndex = None
    hasBeenSaved = None
    flag = None

    def clearPropSeqArray(self):
        self.propSeqCount = 0
        _0 = True
        i = 0
        while True:
            if _0 is True:
                _0 = False
            else:
                i += 1
            if not (i < self.parentClass.getNumProperties()):
                break
            self.prpSequence[i] = 0

    def __init__(self, parClass):
        super(DSSObjectImpl, self)(parClass.getName())
        self.objType = 0
        self.propSeqCount = 0
        self.parentClass = parClass
        self.propertyValue = [None] * self.parentClass.getNumProperties()
        # init'd to zero when allocated
        self.prpSequence = [None] * self.parentClass.getNumProperties()
        self.hasBeenSaved = False

    def finalize(self):
        _0 = True
        i = 0
        while True:
            if _0 is True:
                _0 = False
            else:
                i += 1
            if not (i < self.parentClass.getNumProperties()):
                break
            self.propertyValue[i] = ''
        self.propertyValue = [None] * 0
        self.prpSequence = [None] * 0
        super(DSSObjectImpl, self).finalize()

    def dumpProperties(self, f, complete):
        print 
        print 'new ' + self.getDSSClassName() + '.' + self.getName()

    def edit(self):
        """Allow calls to edit from object itself."""
        self.parentClass.setActiveElement(self.classIndex)
        return self.parentClass.edit()

    def getPropertyValue(self, index):
        """Use DSSClass.propertyIndex() to get index by name."""
        # default behavior for all DSS objects
        return self.propertyValue[index]

    def initPropertyValues(self, arrayOffset):
        self.propertyValue[arrayOffset] = ''
        # clear propertySequence array after initialization
        self.clearPropSeqArray()

    def saveWrite(self, f):
        # Write only properties that were explicitly set in the
        # final order they were actually set.

        iProp = self.getNextPropertySet(0)
        # works on activeDSSObject
        while iProp > 0:
            pc = self.parentClass
            f.print_(' ' + pc.getPropertyName()[pc.getRevPropertyIdxMap()[iProp]])
            f.print_('=' + Utilities.checkForBlanks(self.propertyValue[iProp]))
            iProp = self.getNextPropertySet(iProp)

    def getNextPropertySet(self, idx):
        """Find next larger property sequence number
        return 0 if none found
        """
        smallest = 9999999
        # some big number
        result = 0
        if idx > 0:
            idx = self.prpSequence[idx]
        _0 = True
        i = 0
        while True:
            if _0 is True:
                _0 = False
            else:
                i += 1
            if not (i < self.parentClass.getNumProperties()):
                break
            if self.prpSequence[i] > idx:
                if self.prpSequence[i] < smallest:
                    smallest = self.prpSequence[i]
                    result = i
        return result

    def setName(self, value):
        # if renamed, then let someone know so hash list can be updated
        if len(self.getLocalName()) > 0:
            self.parentClass.setElementNamesOutOfSynch(True)
        self.setLocalName(value)

    def setPropertyValue(self, index, value):
        self.propertyValue[index] = value
        # keep track of the order in which this property was
        # accessed for save command
        self.propSeqCount += 1
        self.prpSequence[index] = self.propSeqCount

    def getName(self):
        return self.getLocalName()

    def getDSSObjType(self):
        return self.objType

    def setDSSObjType(self, type):
        self.objType = type

    def getParentClass(self):
        return self.parentClass

    def setParentClass(self, parent):
        self.parentClass = parent

    def getClassIndex(self):
        return self.classIndex

    def setClassIndex(self, index):
        self.classIndex = index

    def isHasBeenSaved(self):
        return self.hasBeenSaved

    def setHasBeenSaved(self, saved):
        self.hasBeenSaved = saved

    def isFlag(self):
        return self.flag

    def setFlag(self, value):
        # FIXME Protected member in OpenDSS
        self.flag = value

    def getPrpSequence(self):
        return self.prpSequence
